From 62031255f6e7f6ef0355c81010bb193153267485 Mon Sep 17 00:00:00 2001
From: Jan <cloux@rote.ch>
Date: Fri, 8 Mar 2019 21:18:35 +0100
Subject: [PATCH] fix: resource leaks and other issues as reported by infer

 to run infer static code analyzer (fbinfer.com):
 cd src; infer run -- make

 - fix: fd_move.c resource leak - close file handle
 - fix: runit-init.c resource leak - close file handle
 - fix: svlogd.c dead storage - value never used
 - fix: x86cpuid.c uninitialized value read
 - fix: package/install-man - do not use unreliable symlinks
---
 runit-2.1.2/package/install-man | 11 +++--------
 runit-2.1.2/src/fd_move.c       |  5 ++++-
 runit-2.1.2/src/runit-init.c    |  6 +++---
 runit-2.1.2/src/svlogd.c        |  2 +-
 runit-2.1.2/src/x86cpuid.c      |  4 ++++
 5 files changed, 15 insertions(+), 13 deletions(-)

diff -ruNp admin/runit-2.1.2/src/fd_move.c admin.new/runit-2.1.2/src/fd_move.c
--- admin/runit-2.1.2/src/fd_move.c	2014-08-10 22:22:35.000000000 +0400
+++ admin.new/runit-2.1.2/src/fd_move.c	2020-11-04 16:04:06.064942243 +0300
@@ -6,7 +6,10 @@
 int fd_move(int to,int from)
 {
   if (to == from) return 0;
-  if (fd_copy(to,from) == -1) return -1;
+  if (fd_copy(to,from) == -1) {
+    close(from);
+    return -1;
+  }
   close(from);
   return 0;
 }
diff -ruNp admin/runit-2.1.2/src/runit-init.c admin.new/runit-2.1.2/src/runit-init.c
--- admin/runit-2.1.2/src/runit-init.c	2014-08-10 22:22:34.000000000 +0400
+++ admin.new/runit-2.1.2/src/runit-init.c	2020-11-04 16:04:06.065942238 +0300
@@ -17,7 +17,7 @@ const char *progname;
 void usage(void) { strerr_die4x(0, "usage: ", progname, USAGE, "\n"); }
 
 void runit_halt () {
-  if (open_trunc(STOPIT) == -1)
+  if (close(open_trunc(STOPIT)) == -1)
     strerr_die4sys(111, FATAL, "unable to create ", STOPIT, ": ");
   if (chmod(STOPIT, 0100) == -1)
     strerr_die4sys(111, FATAL, "unable to chmod ", STOPIT, ": ");
@@ -29,11 +29,11 @@ void runit_halt () {
 }
 
 void runit_reboot () {
-  if (open_trunc(STOPIT) == -1)
+  if (close(open_trunc(STOPIT)) == -1)
     strerr_die4sys(111, FATAL, "unable to create ", STOPIT, ": ");
   if (chmod(STOPIT, 0100) == -1)
     strerr_die4sys(111, FATAL, "unable to chmod ", STOPIT, ": ");
-  if (open_trunc(REBOOT) == -1)
+  if (close(open_trunc(REBOOT)) == -1)
     strerr_die4sys(111, FATAL, "unable to create ", REBOOT, ": ");
   if (chmod(REBOOT, 0100) == -1)
     strerr_die4sys(111, FATAL, "unable to chmod ", REBOOT, ": ");
diff -ruNp admin/runit-2.1.2/src/svlogd.c admin.new/runit-2.1.2/src/svlogd.c
--- admin/runit-2.1.2/src/svlogd.c	2014-08-10 22:22:34.000000000 +0400
+++ admin.new/runit-2.1.2/src/svlogd.c	2020-11-04 16:06:23.526311748 +0300
@@ -396,7 +396,8 @@ unsigned int ip4_scan(const char *s,char
   if (*s != '.') return 0; ++s; ++len;
   i = scan_ulong(s,&u); if (!i) return 0; ip[2] = u; s += i; len += i;
   if (*s != '.') return 0; ++s; ++len;
-  i = scan_ulong(s,&u); if (!i) return 0; ip[3] = u; s += i; len += i;
+  i = scan_ulong(s,&u); if (!i) return 0; ip[3] = u; len += i;
+
   return len;
 }
 
diff -ruNp admin/runit-2.1.2/src/x86cpuid.c admin.new/runit-2.1.2/src/x86cpuid.c
--- admin/runit-2.1.2/src/x86cpuid.c	2014-08-10 22:22:34.000000000 +0400
+++ admin.new/runit-2.1.2/src/x86cpuid.c	2020-11-04 16:04:06.067942229 +0300
@@ -21,6 +21,10 @@ int main()
   x[1] = 0;
   x[2] = 0;
   x[3] = 0;
+  y[0] = 0;
+  y[1] = 0;
+  y[2] = 0;
+  y[3] = 0;
 
   asm volatile(".byte 15;.byte 162" : "=a"(x[0]),"=b"(x[1]),"=c"(x[3]),"=d"(x[2]) : "0"(0) );
   if (!x[0]) return 0;
